### PR TITLE
many: check that users of BaseTest don't forget to consume cleanups

### PR DESCRIPTION
**Pull request by pedronis**
make their order of execution also LIFO which is more expected

fix places that were wrong, also cleanup some of devicestate
suites to use AddCleanup more and not need specific TearDownTest

### CODE REVIEW

File: overlord/devicestate/handlers_test.go [hunk 1]
```diff
@@ -436,27 +436,29 @@ func (s *deviceMgrSuite) TestDoPrepareRemodeling(c *C) {
...
-	s.BaseTest.SetUpTest(c)
-
-	s.AddCleanup(snapdenv.MockPreseeding(preseed))
+	r := snapdenv.MockPreseeding(preseed)
 
 	// preseed mode helper needs to be mocked before setting up
 	// deviceMgrBaseSuite due to device Manager init.
 	s.deviceMgrBaseSuite.SetUpTest(c)
 
+	// can use cleanup only after having called base SetUpTest
```
Comment (anonymouse64) [inline overlord/devicestate/handlers_test.go:450]: thanks for the comment here

**Review by mvo5 — APPROVED** [review]
Looks good - one question/suggestion inline.

**Review by bboozzoo — APPROVED** [review]

File: cmd/snap-repair/main_test.go [hunk 2]
```diff
@@ -76,6 +76,10 @@ func (r *repairSuite) SetUpTest(c *C) {
 	r.AddCleanup(func() { dirs.SetRootDir("/") })
 }
 
+func (r *repairSuite) TearDownTest(c *C) {
+	r.BaseTest.TearDownTest(c)
```
[the remark this answers is not in the document]
Comment (pedronis) [reply cmd/snap-repair/main_test.go:80]: no, you are right, I was somehow confused (I thought this was using only the Suite methods from that)

File: overlord/devicestate/devicestate_test.go [hunk 3]
```diff
@@ -208,25 +211,15 @@ func (s *deviceMgrBaseSuite) SetUpTest(c *C) {
...
-	s.ancillary = nil
-	s.state.Lock()
-	assertstate.ReplaceDB(s.state, nil)
-	s.state.Unlock()
-	bootloader.Force(nil)
-	dirs.SetRootDir("")
-	s.restoreGenericClassicMod()
-	s.restoreOnClassic()
-	s.restoreSanitize()
-	s.restoreCloudInitStatusRestore()
```
Comment (zyga) [inline overlord/devicestate/devicestate_test.go:227]: This is not called anymore, is this intentional?

File: testutil/base.go [hunk 4]
```diff
@@ -31,15 +31,19 @@ type BaseTest struct {
 
 // SetUpTest prepares the cleanup
 func (s *BaseTest) SetUpTest(c *check.C) {
-	s.cleanupHandlers = nil
+	if len(s.cleanupHandlers) != 0 {
+		panic("BaseTest cleanup handlers were not consumed before a new test start, missing BaseTest.TearDownTest call?")
```
[the remark this answers is not in the document]
Comment (zyga) [reply testutil/base.go:35]: Yeah, this is very useful, thank you Samuele!

File: overlord/devicestate/devicestate_test.go (hunk 3 above)
Comment (pedronis) [reply overlord/devicestate/devicestate_test.go:227]: yes, everything was moved to use AddCleanup from SetUpTest and in this case the BaseTest.TearDownTest is called for us so nothing to override/do

**Pull request merged**